src/compiler/checker.ts

```diff
@@ -7928,22 +7928,17 @@ namespace ts {
             const numTypeArguments = length(typeArguments);
             if (isJavaScriptImplicitAny || (numTypeArguments >= minTypeArgumentCount && numTypeArguments <= numTypeParameters)) {
                 const result = typeArguments ? typeArguments.slice() : [];
-
-                // Map an unsatisfied type parameter with a default type.
-                // If a type parameter does not have a default type, or if the default type
-                // is a forward reference, the empty object type is used.
-                const baseDefaultType = getDefaultTypeArgumentType(isJavaScriptImplicitAny);
-                const circularityMapper = createTypeMapper(typeParameters!, map(typeParameters!, () => baseDefaultType));
+                // Map invalid forward references in default types to the error type
                 for (let i = numTypeArguments; i < numTypeParameters; i++) {
-                    result[i] = instantiateType(getConstraintFromTypeParameter(typeParameters![i]) || baseDefaultType, circularityMapper);
+                    result[i] = errorType;

```

Comment:
This is effectively going back to the [behavior in the original PR](https://github.com/Microsoft/TypeScript/pull/13487/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR5327). It looks like the circularity was introduced in #28222 and #28423. Though, prior to this we were just defaulting to `{}` in TypeScript (`any` in JavaScript).